### PR TITLE
fix(rust_extractor): process env vars and compiler args passed to wrapper

### LINKED ISSUE #5011: Rust extractor doesn't support pulling in compiler wrapper arguments
The Bazel Rust rules utilize a wrapper binary to configure the build environment before calling the Rust compiler. This brings in environment variables and other information output by build scripts. Currently, the Rust extractor does not replicate this behavior, which can cause extraction to fail when a compilation target requires some of those environment variables or linking arguments to compile.

### PR DESCRIPTION
**Pull request by wcalandro**
Fixes #5011

### CODE REVIEW

File: kythe/rust/extractor/src/bin/save_analysis.rs [hunk 1]
```diff
@@ -55,5 +57,53 @@ fn generate_arguments(arguments: Vec<String>, output_dir: &Path) -> Result<Vec<S
         output_dir.to_str().ok_or_else(|| anyhow!("Couldn't convert temporary path to string"))?;
     rustc_arguments[outdir_position] = format!("--out-dir={}", outdir_str);
 
+    // Check if there is a file containing environment variables
+    let env_file_position = arguments.iter().position(|arg| arg.contains("--env-file"));
```
Comment (shahms) [inline kythe/rust/extractor/src/bin/save_analysis.rs:61]: `arg.starts_with` or `==` ?

File: kythe/rust/extractor/src/bin/save_analysis.rs [hunk 2]
```diff
@@ -55,5 +57,53 @@ fn generate_arguments(arguments: Vec<String>, output_dir: &Path) -> Result<Vec<S
...
+    if let Some(position) = env_file_position {
+        let env_file = &arguments[position + 1];
+        let path = Path::new(env_file);
+        process_env_file(path).context("Failed to process wrapper environment file")?;
+    }
+
+    // Check if there is a file containing extra compiler arguments
+    let arg_file_position = arguments.iter().position(|arg| arg.contains("--arg-file"));
```
Comment (shahms) [inline kythe/rust/extractor/src/bin/save_analysis.rs:69]: ditto

File: kythe/rust/extractor/src/bin/save_analysis.rs [hunk 3]
```diff
@@ -55,5 +57,53 @@ fn generate_arguments(arguments: Vec<String>, output_dir: &Path) -> Result<Vec<S
...
+/// * `path` - The path of the compiler argument file
+fn process_arg_file(path: &Path) -> Result<Vec<String>> {
+    let file = File::open(path)?;
+    let lines = io::BufReader::new(file).lines();
+    let mut arguments = Vec::new();
+    for line in lines {
+        let line_string = line.context("Failed to read line of arg file")?;
+        arguments.push(line_string);
+    }
+    Ok(arguments)
```
Comment (shahms) [inline kythe/rust/extractor/src/bin/save_analysis.rs:108]: FWIW, this is also just:
```
    io::BufReader::new(File::open(path)?)
        .lines()
        .map(|line| line.context("..."))
        .collect()
```

File: kythe/rust/extractor/src/bin/save_analysis.rs (hunk 1 above)
Comment (wcalandro) [reply kythe/rust/extractor/src/bin/save_analysis.rs:61]: Oh yes this should be ==

File: kythe/rust/extractor/src/bin/save_analysis.rs (hunk 3 above)
Comment (wcalandro) [reply kythe/rust/extractor/src/bin/save_analysis.rs:108]: Oh I actually never knew that a question mark could be used inline. However, the iterator returns an Option so I don't believe that map would work correctly.

Comment (wcalandro) [reply kythe/rust/extractor/src/bin/save_analysis.rs:104]: I stand corrected

File: kythe/rust/extractor/src/bin/save_analysis.rs [hunk 4]
```diff
@@ -55,5 +57,47 @@ fn generate_arguments(arguments: Vec<String>, output_dir: &Path) -> Result<Vec<S
         output_dir.to_str().ok_or_else(|| anyhow!("Couldn't convert temporary path to string"))?;
     rustc_arguments[outdir_position] = format!("--out-dir={}", outdir_str);
 
+    // Check if there is a file containing environment variables
+    let env_file_position = arguments.iter().position(|arg| arg == "--env-file");
+    if let Some(position) = env_file_position {
+        let env_file = &arguments[position + 1];
```
Comment (shahms) [inline kythe/rust/extractor/src/bin/save_analysis.rs:63]: This (and below) assume the arguments are well-formed such that there is an argument value at `position + 1`.  I suspect using `arguments.windows(2)` may be more robust?

```
for window in arguments.windows(2) {
  if let [arg, value] = window {
    if arg == "--env-file" {
      process_env_file(Path::new(value)).context("....")?;
      break;
    }
  }
}
```

Comment (wcalandro) [reply kythe/rust/extractor/src/bin/save_analysis.rs:63]: This operation is controlled by the Rust rules, and these arguments are being passed before the arguments that are passed to the Rust compiler, so there is a guarantee that there will at least be an argument after the flag, so we will not have an out-of-bounds read

**Review by shahms — APPROVED** [review]

**Pull request merged**